### ISSUE TITLE
Two recommendations on the use of software

### ISSUE THREAD

**Issue by iheshun** [lwouis/alt-tab-macos#337]
I'm sorry, I'm a Chinese person, so I'm using translation software to write this paragraph, if there is a problem with the expression, please understand.

Currently using the software you provided and it's great. But in use, I have 2 small comments that I hope you will consider.

1、Does activating (command+tap) the "panel" allow setting shortcut keys? This allows me to set shortcuts on the mouse and activate the panel with one click.
2、After the panel is activated, the current App is sorted according to the order of the window, whether you can support sorting by application category. For example, 1(word)3(word),2(Powerpoint),3(Excel),3(Evernote),4(Facetime)

Finally thanks again for making such great software and good luck!

Translated with www.DeepL.com/Translator (free version)

**Comment by lwouis** [comment]
Hi @iheshun! Thanks for sharing this feedback!

I will try to answer your suggestions:

> 1. Does activating (command+tap) the "panel" allow setting shortcut keys

You can change the shortcuts easily in the preferences panel:

![image](https://user-images.githubusercontent.com/106195/81789298-baa35a00-953e-11ea-8ea5-75e42cce7cc9.png)

There you can configure every shortcut:

![image](https://user-images.githubusercontent.com/106195/81789339-cd1d9380-953e-11ea-9b9d-7903a9a90f91.png)

If you want to trigger the shortcuts using the mouse, most mouse manufacturers let you assign keyboard keys to a mouse click. On my Logicool mouse for example, I set it like this:

![image](https://user-images.githubusercontent.com/106195/81789480-02c27c80-953f-11ea-8d53-0436d0d4a332.png)

For example, you would set your mouse click4 to send `alt-tab`, then you would set AltTab to trigger on `alt-tab`. You can change the keyboard shortcut of course.

> 2. After the panel is activated, the current App is sorted according to the order of the window, whether you can support sorting by application category. For example, 1(word)3(word),2(Powerpoint),3(Excel),3(Evernote),4(Facetime)

It is an interesting idea, but I think it is a bit complex and may not be as useful as it seems. Let's say I have this today:

Word1, Powerpoint, Excel, Evernote, Facetime, Word2, Word3, Word4

If we group by app, then you would see this:

Word1, Word2, Word3, Word4, Powerpoint, Excel, Evernote, Facetime

So even though you haven't opened Word2, Word3, Word4 maybe in a long time, they are taking up all the prime real-estate on the UI, just because you recently opened Word1.

That being said, maybe other people would want that feature, so I'm going to keep this ticket open, and rename it to reflect your second suggestion. If it gets enough upvotes, then I'll probably implement it. Another option is that you, or someone who wants that feature submits a pull-request. I would gladly merge it as it would a preference, so would not impact people who don't want it, but would help people who want it.

**Comment by iheshun** [comment]
Hi ！

Thank you so much for your reply, I can see that you are a very enthusiastic person.

First of all, I regret that, as a questioner, one should first identify one's question carefully. The first question I misdescribed and caused you unnecessary trouble.

I tried to type "option+tab", but I ended up typing "command+tab".

Regarding the first question, I found the cause of the problem, I couldn't activate "option+tab" with the mouse because I used an external keyboard and modified the modifier keys.  picture1
So when I was setting the mouse shortcut, it changed to  picture2  , and the correct one should be   picture3  . Now the first problem has been solved.

Regarding the second question, the starting point is to improve the efficiency of software positioning（定位：fixed position）, we usually switch between similar applications (COMMAND+C, COMMAND+V) and similar software is easier to find when presented together. Here I found two pictures for comparison.    Picture4    and    picture5

This problem can also be solved if you can set the applications to be in window order.
Like this: 3 (current window), 1, 2, 4, 5.
Because I can select a certain application to open in a specific window, like in     picture6    .

Like you said, this feature may not be very important and doesn't require you to spend much time thinking about it.

Finally thanks again for your reply and good luck.


  Picture1



  Picture2



  Picture3



  Picture4



  Picture5



  Picture6


Harlan
------
heshun@outlook.com
在 2020年5月13日 +0800 PM4:36，lwouis <notifications@github.com>，写道：

**Comment by iheshun** [comment]
Image 6 was not sent successfully, I tried to send it again.
"Assigned to this desktop"



________________________________
发件人: he shun <heshun@outlook.com>
发送时间: 2020年5月13日 10:16
收件人: lwouis/alt-tab-macos <reply@reply.github.com>
主题: Re: [lwouis/alt-tab-macos] Two recommendations on the use of software (#337)

Hi ！

Thank you so much for your reply, I can see that you are a very enthusiastic person.

First of all, I regret that, as a questioner, one should first identify one's question carefully. The first question I misdescribed and caused you unnecessary trouble.

I tried to type "option+tab", but I ended up typing "command+tab".

Regarding the first question, I found the cause of the problem, I couldn't activate "option+tab" with the mouse because I used an external keyboard and modified the modifier keys.  picture1
So when I was setting the mouse shortcut, it changed to  picture2  , and the correct one should be   picture3  . Now the first problem has been solved.

Regarding the second question, the starting point is to improve the efficiency of software positioning（定位：fixed position）, we usually switch between similar applications (COMMAND+C, COMMAND+V) and similar software is easier to find when presented together. Here I found two pictures for comparison.    Picture4    and    picture5

This problem can also be solved if you can set the applications to be in window order.
Like this: 3 (current window), 1, 2, 4, 5.
Because I can select a certain application to open in a specific window, like in     picture6    .

Like you said, this feature may not be very important and doesn't require you to spend much time thinking about it.

Finally thanks again for your reply and good luck.

[cid:26E29A9A6EF047EE8CA49D3A696A4586]
  Picture1


[cid:DAD9B9FD79AA4F0DABB0E1FFCC15F29A]
  Picture2


[cid:E6FB7F40F7B6427683D31CCCD018D898]
  Picture3


[cid:1D2AACC7025B443699AE3C3F2E94A5D8]
  Picture4


[cid:919AE03804254796907A0B3E8DAB9A67]
  Picture5


[cid:EA9D9CEC10A741EE9ECE44CC709F6E2F]
  Picture6


Harlan
------
heshun@outlook.com
在 2020年5月13日 +0800 PM4:36，lwouis <notifications@github.com>，写道：

Hi @iheshun<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fiheshun&data=02%7C01%7C%7C71e876aa512c44852bab08d7f718b1ac%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637249557733315836&sdata=ubHqVfIIRA9oNbLhQge8a6NYvjDC9MvP7gAuizl2CQM%3D&reserved=0>! Thanks for sharing this feedback!

I will try to answer your suggestions:

  1.  Does activating (command+tap) the "panel" allow setting shortcut keys

You can change the shortcuts easily in the preferences panel:

[image]<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F106195%2F81789298-baa35a00-953e-11ea-8ea5-75e42cce7cc9.png&data=02%7C01%7C%7C71e876aa512c44852bab08d7f718b1ac%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637249557733329235&sdata=9jkV4ZGBlVMGu8jBJw%2B10AMCH2kN556YPFsb4FQQYS4%3D&reserved=0>

There you can configure every shortcut:

[image]<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F106195%2F81789339-cd1d9380-953e-11ea-9b9d-7903a9a90f91.png&data=02%7C01%7C%7C71e876aa512c44852bab08d7f718b1ac%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637249557733335830&sdata=B3qC4LybcOithpj2N%2FTRtGrcJWNuJK6FDzuYhPSyZMc%3D&reserved=0>

If you want to trigger the shortcuts using the mouse, most mouse manufacturers let you assign keyboard keys to a mouse click. On my Logicool mouse for example, I set it like this:

[image]<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F106195%2F81789480-02c27c80-953f-11ea-8d53-0436d0d4a332.png&data=02%7C01%7C%7C71e876aa512c44852bab08d7f718b1ac%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637249557733345823&sdata=Wppiui0KWTFZobgWtPIURe1xdQtQH9x96wUFXHXZSFM%3D&reserved=0>

For example, you would set your mouse click4 to send alt-tab, then you would set AltTab to trigger on alt-tab. You can change the keyboard shortcut of course.

  1.  After the panel is activated, the current App is sorted according to the order of the window, whether you can support sorting by application category. For example, 1(word)3(word),2(Powerpoint),3(Excel),3(Evernote),4(Facetime)

It is an interesting idea, but I think it is a bit complex and may not be as useful as it seems. Let's say I have this today:

Word1, Powerpoint, Excel, Evernote, Facetime, Word2, Word3, Word4

If we group by app, then you would see this:

Word1, Word2, Word3, Word4, Powerpoint, Excel, Evernote, Facetime

So even though you haven't opened Word2, Word3, Word4 maybe in a long time, they are taking up all the prime real-estate on the UI, just because you recently opened Word1.

That being said, maybe other people would want that feature, so I'm going to keep this ticket open, and rename it to reflect your second suggestion. If it gets enough upvotes, then I'll probably implement it. Another option is that you, or someone who wants that feature submits a pull-request. I would gladly merge it as it would a preference, so would not impact people who don't want it, but would help people who want it.

―

**Comment by lwouis** [comment]
Actually, all the pictures in your post are not showing. I think there was an issue with you posting them using email instead of the github.com website.

**Comment by iheshun** [comment]
The email client I used was spark; there was indeed a problem.

The first problem has been solved, the reasons are:
1. I used an external keyboard;【Pic 1】
2. I modified the modifier key;【Pic 2】
![000](https://user-images.githubusercontent.com/62940565/81952666-56b18c00-9639-11ea-8e0f-051e7e091c1f.png)
【Pic 1】

![002](https://user-images.githubusercontent.com/62940565/81953030-c45db800-9639-11ea-9169-7a326a6a5951.png)
【Pic 2】

So I wanted to set option + tab, but in the end it became command + tab, and the Logitech panel display in Chinese is also wrong（tab translates to "标签"）【pic 3】, causing me to find no problems.
![007](https://user-images.githubusercontent.com/62940565/81954073-ff142000-963a-11ea-9c2f-e60a74dc63aa.png)

【Pic 3】

The second question, I mainly consider the situation of frequently copying and pasting content in the same application. Here I use 2 pictures as examples.【Pic 4】and 【Pic 5】
![008](https://user-images.githubusercontent.com/62940565/81953552-6382af80-963a-11ea-8a53-c38fc2d3b903.png)
【Pic 4】

![009](https://user-images.githubusercontent.com/62940565/81953588-709f9e80-963a-11ea-98da-fec3e8349408.png)
【Pic 5】

In fact, I thought about a compromise method, as long as the application is allowed to sort according to the window serial number. 
Like this: 3 (current window), 1, 2, 4, 5.

Because I can set the application to appear in only one window.【Pic 6】
![003](https://user-images.githubusercontent.com/62940565/81953671-8e6d0380-963a-11ea-935f-9a3261844d1c.png)
【Pic 6】

**Comment by lwouis** [comment]
AltTab is a port of Windows 10. As such, it shows windows in chronological order. While I believe it to be the best general approach for productivity, I'm open to introducing other workflows to the app. People have suggested direct access for instance (#210).

The issue however with adding more workflows into the same tool, is the explosion of preferences. The tool already has a large number of preferences because I find myself wanting to let people have it their way, and be empowered. The downside of that is that the tool is becoming unwelcoming/overwhelming, especially to new and more casual users.

Let's imagine we add a new preference to group the windows by app. That's yet one more preference. Now in your example (Pic 5), you say that the windows for each app could be ordered based on the screen they are on. Some people may prefer that they be in chronological order, instead of based on which screen. Actually if you have multiple windows of the same app on the same screen, you need to sort them anyway, and chronological seems the way to go. My point being that we are already talking about adding 2 new preferences here.

Another issue is that I decided our of principles to have 0 telemetry in the app, website, etc. This means that I have 0 statistics on users preferences, usage pattern, etc, outside of the active people like you who take the time to discuss on the Issues here. I think everyone would agree that adding a preference to accommodate 1 person is not the best idea. Preferences should be useful to a bunch of people. However as I said there is no way to know even if current preferences in the app are of any use.

Really, I'm not sure how to pilot this project forwards without data. I all falls down to my judgment in the end since I'm the main contributor. It's not necessarily a bad thing btw to not follow data. Data is why fast food in the US is so big. They optimize demand and human feedback for caloric-rich food, and now they have a terrible obesity problem. I rather like the idea of not always optimizing for real-world usage as people's habits may not be the direction we should reinforce.

I'll keep this ticket open as a discussion point. I think I'll open a new ticket to kind of review the current preferences and features of the app, and discuss the project direction there.

**Comment by BBartosz** [comment]
I think OP is talking about behaviour which you might find in Ubuntu or other gnome UI's

![image](https://user-images.githubusercontent.com/5239728/83848351-f2fe1880-a737-11ea-90bc-ecff2d475ae5.png)

I find it very useful, would be a great feature. Anyway this plugin save my life in macos, amazing job 👏

**Comment by lwouis** [comment]
I was thinking about this, and observed that this is creating a hierarchy. AltTab currently shows all the information on first interaction. All windows are shown, annotated with which app, Space, there are of, with title, minimized/hidden state, etc. Here you are talking about showing a first organization layer, then drilling down. Note that this concept could be down for any characteristic of the windows. On the screenshot, the grouping is done by app, but you could imagine presenting Spaces, screens, etc.

The way I see it is that it's the same information, presenting in a different layout. Other layouts have been requested, such as #384. To be honest, I would like to provide the option. Having a Layout preference where you select between:

* All windows shown (classic)
* Group by app
* Group by Space
* Group by Screen

The issue right now is time. Making the original layout took me hundreds of hours. It's quite hard, as it turns out, to do layout when you support dozens of preferences that affect it, different screen sizes and ratios, scrolling when there are too many windows (mouse wheel and trackpad need separate support), drag-and-drop on the windows, mouse-hover, click of the windows and out of the ui to close it, and importantly show the ui within a few frames of the shortcut being frame. The current UI is seemingly simple but represent a pretty insane amount of time and knowledge.

Anyway PRs welcome, as always ;p

**Comment by dvcrn** [comment]
I came here from the https://github.com/lwouis/alt-tab-macos/issues/532 feature request. 

I want to give my +1 for a feature that allows to restore the existing CMD-TAB functionality to switch based on applications, and not windows. 

So why not just use the built-in thing then? Well it doesn't look as nice, but it can also not be restricted to the current space. The built-in switcher will always grab all apps across all spaces. 

It could be as simple as adding an option here that specifies whether to show windows or apps. 

<img width="360" alt="Screen Shot 2021-06-23 at 13 05 17" src="https://user-images.githubusercontent.com/688326/123033991-aa3cfe80-d423-11eb-9e88-9888492b5352.png">

So I can see 3 main patterns: 

- Switch across all apps
- Switch across all windows
- Switch across current apps windows

**Comment by wfaulk** [comment]
I just wanted to mention that HyperSwitch does (did?) this as well. It somehow tied into the normal Cmd-Tab UI and showed the list of windows for the highlighted app as a separate subwindow:

![Hyperswitch](https://user-images.githubusercontent.com/205470/127058752-20c63272-8026-4e4e-8ffc-717f3707814b.png)

**Comment by ggedde** [comment]
Hi @lwouis,
Thank you and all contributors that have worked on this.
It is really awesome and I appreciate it.

I understanding that drilling down is the main question right now, but honestly I just need a "Group by Apps" option.
I don't need to show or drill down multiple windows from the same apps.  I would just prefer that when selecting a grouped app that it just focuses on the last opened window from that app. Most of my apps have their own keyboard command for switching between windows within them. I can see some people wanting a number on the grouped apps that show how many there are, but I personally don't need that. Just a simple a checkbox for "Group by Apps". It would be great to get that added in then keep working with the community on ideas for drilling down and then changing this feature once that has been decided, etc.

**Comment by ejabu** [comment]
Hi, @lwouis @ggedde @iheshun 

I looked into the codebase, and I assume it has something to do with [this function](
https://github.com/lwouis/alt-tab-macos/blob/f5fc5ba8131fb65a2e54cf5c1795626b5118c1c4/src/logic/Windows.swift#L10
)  called `reorderList`

```swift
static func reorderList() {
        list.sort {
            if let bool = sortByBooleanAttribute($0.isWindowlessApp, $1.isWindowlessApp) {
                return bool
            }
            if Preferences.showHiddenWindows[App.app.shortcutIndex] == .showAtTheEnd,
               let bool = sortByBooleanAttribute($0.isHidden, $1.isHidden) {
                return bool
            }
            if Preferences.showMinimizedWindows[App.app.shortcutIndex] == .showAtTheEnd,
               let bool = sortByBooleanAttribute($0.isMinimized, $1.isMinimized) {
                return bool
            }
            return $0.lastFocusOrder < $1.lastFocusOrder
        }
    }
 ```
 
 
![image832](https://user-images.githubusercontent.com/2019630/147812879-9e118447-23b9-41bc-8f92-d4dfc2b72128.png)

 
 My simple solution is first, we may group these Visual Studio Code adjacent to its counterpart and also group Google Chrome next to another two windows.
 
 we can sort it by the Suffix of Window Title
 
 ```
 - Visual Studio Code
 ```


```
- Google Chrome
```

**Comment by lwouis** [comment]
@ejabu there is no need to parse the app name from the title. You can access the app name directly from `$0.application.runningApplication.localizedName`. However, you don't want to group/match by localized name. You should do it based on the bundleIdentifier which uniquely identifies an app on macOS, so `$0.application.runningApplication.bundleIdentifier`.

Note that the sorting of windows based on app or space is a very small part of this ticket. Most of the work lies in thinking about the UX that would be added, and these topics here:

* https://github.com/lwouis/alt-tab-macos/issues/337#issuecomment-673304578
* https://github.com/lwouis/alt-tab-macos/issues/337#issuecomment-930288519

There are many workflows requested by various users here. It would be nice to find a way to combine them, or find a UX which can satisfy various users even though they are currently asking for different things. Some asked for a drill-down UI, some just want grouping without any way to drill-drown, some want a main/details split-view [like this](https://github.com/lwouis/alt-tab-macos/issues/337#issuecomment-887024775), etc.

There is some UX/workflow thinking to do here. Then there is work specifying in more details how it would work and how it would integrate with all other features of the app. Then it's implementation time, of which I think most of the work will be on a new layout system.

**Comment by alonpr** [comment]
@lwouis , I believe that a "simple" optional sorting by the app name, as suggested by @ejabu , can be helpful for many users, especially when working with multiple windows of the same app.

**Comment by universvm** [comment]
Are there any updates on this by any chance?

**Comment by vanchKong** [comment]
nothing changed  :cry:
This feature makes AltTab perfect.

**Comment by XOR-op** [comment]
Hello @lwouis, since #1801 I have created a fork to fulfill my own need. That works well to me for months. Therefore, I want to hear from your opinion and if you think it's ok, I will open a pull request to address current issue.

Here is what I added: 

<img width="560" alt="alttab" src="https://user-images.githubusercontent.com/17672363/213380508-687a3833-dcec-4d52-91ce-f251af7f0447.png">

I create another option "Group By", of which the currently available options are "Disable" (classic AltTab ordering) and "By space" (windows are grouped by space, numerically increasing, and the ordering of applications inside the same space follows the classic). If I had time I may add other options like "By app", or it can be added in other PRs. 

Here I want to see your response including: 1. Is the design ok? 2. Is the naming ok, or you think there are better names?

**Comment by lwouis** [comment]
Hi @XOR-op,

It seems to me that you implemented sorting, not grouping. Grouping would manifest the group visually. It could be a line surrounding the items from a group, a background color, etc.

Could you share what the UI looks like? It would help understand what your new preference does exactly.

This ticket here is a bit ambitious. It aims to have groups and drill-down. So users have "levels" like folders, and can see one level, then go down another. The goal being to not be overwhelmed by a large number of windows

**Comment by universvm** [comment]
imo having sorting by application is already a step forward but I'd agree it'd be better to see what the feature looks like

**Comment by XOR-op** [comment]
> Hi @XOR-op,
> 
> It seems to me that you implemented sorting, not grouping. Grouping would manifest the group visually. It could be a line surrounding the items from a group, a background color, etc.
> 
> Could you share what the UI looks like? It would help understand what your new preference does exactly.
> 
> This ticket here is a bit ambitious. It aims to have groups and drill-down. So users have "levels" like folders, and can see one level, then go down another. The goal being to not be overwhelmed by a large number of windows

The unsorted version:
<img width="1000" alt="unsorted" src="https://user-images.githubusercontent.com/17672363/213441188-c055ea0e-92ea-4f5f-bb2d-83e338703e4c.png">

The sorted version:
<img width="1000" alt="sorted" src="https://user-images.githubusercontent.com/17672363/213441206-ee9bf1b8-b224-4aa9-9167-4b0c28b645f8.png">

This is a purely sorting without any touching with UI. And I also saw #1936 but that PR has been inactive for long. I'm not sure how to contribute is better but I think at least a similar feature should be merged into mainstream. Lots of users are waiting for this.

**Comment by lwouis** [comment]
Let's please keep in topic, and hold conversations about sorting in the [sorting ticket](https://github.com/lwouis/alt-tab-macos/issues/515), or in the [sorting PR](https://github.com/lwouis/alt-tab-macos/pull/1936).

This ticket here is about grouping, drill-down, hierarchy. Scroll up to see the kind of ideas people discuss here.

**Comment by universvm** [comment]
Not related to alt-tab but bettertouchtool released something similar to alt-tab but with this feature and more in case anyone is looking for a solution immediately.

https://updates.folivora.ai/window_switcher.png

**Comment by ZHDI-1** [comment]
I think things like ubuntu do is not really great, user may dont want a extra large icon to fill space and crowding the space of windows, here might has a better way to do it:
```
_______________________________________________________
| small icon with name of application
| ______________
| |              |
| |              |         
| |______________|.....
|here will be a line split space to another application
|another icon and name of another application
| ______________
| |              |
| |              |         
| |______________|.....
| ______________________________________________________
```
we will use `<Tab>` or `<`>` to switch application and another to switch window, arrow keys will support to switch windows and here will two situations:
1. windows of an application only occupie one line, down arrow key will just jump to other application
2. windows of an application occupie more than one line, and down arrow key will jump to another line of application, utils the next line is next application

in summary, arrow keys will only support windows jump unless the next one is another application's.
there may be several problems here, like if a application only has one or two window, then we may should let these things dont take up an entire line...

and, this solution will does not break current layout too much, it should take less time.
I want to hear your opinion!
@lwouis

**Comment by whoisdecoy** [comment]
#2833 I think mimicking native Mac OS and grouping all windows to once icon is a key feature. showing all windows when selecting one icon. I would prefer that as an option over drilling down or having them all separate.

**Comment by azhongzl** [comment]
> #3040 I was able to make the change to the source code to show only one window of each application, in the same way that the native macOS shortcut ali+tab works. Thank you very much.
> 
> ![Screenshot 2023-12-09 at 19 57 50](https://private-user-images.githubusercontent.com/9993158/289316734-6d6c827e-d9cf-46af-b8cb-1f3fe5abc2b3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDIxOTc2NzMsIm5iZiI6MTcwMjE5NzM3MywicGF0aCI6Ii85OTkzMTU4LzI4OTMxNjczNC02ZDZjODI3ZS1kOWNmLTQ2YWYtYjhjYi0xZjNmZTVhYmMyYjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMTBUMDgzNjEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2M0MTM3ZmMzMGJiMTI4ODk2OTZlYTRjYWUzMDdhODQyNzRlODFmYmIyOTMwNmVlZTExNzhlNTg0MTBlYjg5MyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.iooRMcbsY3YOccZly_1LkivFLSCeU_01Q0HKIJniOOM)

Could you please put this to the release so that others can use as well?

**Comment by lwouis** [comment]
Hi @anonymous-si,

Thank you for sharing your work.

I believe, from the screenshot your shared, that you did something quite different from the ambitions of this ticket here. Here is [what we are trying to discuss here](https://github.com/lwouis/alt-tab-macos/issues/337#issuecomment-673304578). There is grouping by different characteristics, and there is navigation within an app (i.e. drilldown).

It seems that you went with a simplified approach where each app only shows 1 window. I'm not sure how many users would use this.

Thank you

**Comment by vip9** [comment]
> Hi @anonymous-si,
> 

> It seems that you went with a simplified approach where each app only shows 1 window. I'm not sure how many users would use this.

If it would be the latest used window, I would use it. To switch to oldests windows from app, there's still cmd+`.

**Comment by azhongzl** [comment]
> Hi @anonymous-si,
> 
> Thank you for sharing your work.
> 
> I believe, from the screenshot your shared, that you did something quite different from the ambitions of this ticket here. Here is [what we are trying to discuss here](https://github.com/lwouis/alt-tab-macos/issues/337#issuecomment-673304578). There is grouping by different characteristics, and there is navigation within an app (i.e. drilldown).
> 
> It seems that you went with a simplified approach where each app only shows 1 window. I'm not sure how many users would use this.
> 
> Thank you

Two ways are accepted from my side:
1. like MacOS mission control->"Group windows by application", it shows all active windows of the same active app in a group, and the latest one on the top
2. just show the latest window of an app, if the first option is hard to implement.

**Comment by lwouis** [comment]
Hi @anonymous-si,

As it stands, it looks like it's not really grouping, but rather showing only 1 window per app/Space/screen. Do you intend on adding the in-group navigation / drilldown? If not, I suggest to rename "Group windows by:" to "Show only first window per:".

Without the in-group navigation, I'm not sure how useful it is to users in that thread. I'm not sure it could be released as it is.

Thank you

**Comment by vurtomatic** [comment]
@anonymous-si I just started a ticket asking about "grouping windows by display".

Or any kind of indicator, such as a badge, to show the window is on a different display would be extremely helpful and quicker when switching windows.

**Comment by erikhuizinga** [comment]
Just dropping by to express my interest in a feature to group the AltTab layout by app. I use 3 displays and 3+ browser windows at all times. While I like the existing feature to only show windows from the active app, I think there indeed is a use case to group windows of all apps. E.g. right now I have 15 open windows of various apps. Switching within my browser I have a shortcut for, but finding a specific app that I used a few screens ago (it's lower on the stack, somewhere in the middle–end), can be hard and require a lot of keystrokes after looking for the correct app window, or even mouse movement (😱 the horror! 🤢).

My personal preferences are:

1. <kbd>⌘</kbd><kbd>⇥</kbd>: My main use case for AltTab is just being able to hit the shorcut and switch back to my previous window, regardless of app. This feature I would never change.
2. <kbd>⌥</kbd><kbd>⇥</kbd>: My secondary use case is the one of this issue: switch to some app further down the (potentially large) stack.
3. <kbd>⌘</kbd><kbd>`</kbd>: My tertiary use case is switching between windows of the current app, which is already a feature.

**Comment by tkunze** [comment]
Love the visual nature of AltTab but miss the one, much simpler feature which makes [Witch](https://manytricks.com/witch/) great: listing windows **in a predictable order**. In particular:

1. Group all windows **by app**, using the order in which apps appear in the Dock.
2. Sort windows within each app by title.

This would massively reduce the cognitive energy needed to find the window one is looking for. No more scanning, simply zero in on the location where that app/window tends to be.

No UI changes. Just one new "Sort apps by" setting in Preferences: Sort apps by: {None, Alphabetical order, Dock order}. The Sort windows by(Alphabetical order) already exists.

**Comment by rhyek** [comment]
It would be great if we could get the "Group windows by application" feature here. I decided to disable the cmd+tab shortcut in AltTab and just use the native mac os switcher in the meantime while keeping cmd+` to switch within same app, but it is not ideal since I have three screens and:
1. The native switcher shows only on the main screen. (Got around this using: `defaults write com.apple.dock appswitcher-all-displays -bool true; killall Dock`)
2. Shows windows from all screens which AltTab has already solved.

I really don't think the current suggestions of being able to group by screen and space are a good idea. If I wanted that I would first switch to the screen or space and press alt+tab or w/e key shortcut there. Also I don't think it's important to "drill down". It'd be much easier to switch to an app with alt+tab, then switch to a different window of it with alt+`.

I think discussions here should be about simplifying the feature requested and then implement that.

**Comment by risalfajar** [comment]
I agree with @rhyek 

>I don't think it's important to "drill down". It'd be much easier to switch to an app with alt+tab, then switch to a different window of it with alt+`.

This is the better approach, trying to add the "drill down" would increase complexity unnecessarily.

>I think discussions here should be about simplifying the feature requested and then implement that.

The feature should be developed as simple as possible, then iterate from there. It has been years, it's too long.

**Comment by shinebayar-g** [comment]
Wholeheartedly agree with Cmd+Tab only switch for apps (not windows). I'm not even sure drill down is actually a good idea.
Cmd+Tab for apps, Cmd+` for windows would be perfect.
